### ISSUE TITLE
Parser error with "continued" LD/ST symbols

### ISSUE THREAD

**Issue by MikeMiller567** [ARM-software/tarmac-trace-utilities#15]
I'm hitting a problem with a newer emitter splitting memory accesses (LD and ST records) across multiple lines, but only providing the LD  or ST symbol for the first memory location. It looks like this:

```
  947020259 ps  ES  (000000008000d1fc:a90d8be1) O el3t_rt:        STP      x1,x2,[sp,#0xd8] 
                    ST 000000009884cfb0  ........ ........ 00000000 000000fe  NS:000000009884cfb0   NM ISH IWBRWA
                       000000009884cfa0  00000000 00000007 ........ ........  NS:000000009884cfa0   NM ISH IWBRWA
```

The parser is unable to parse the second half of the operations because it currently expects the LD or ST symbol on each line, so I get error/warnings like:

```
tarmac.log:3199: unknown Tarmac event type '000000009884cfa0'
```

Would it be reasonable to refactor/rework the parser to assume that a line starting with an address-like value is a continuation of an earlier LD/ST operation or is this something to deal with in the emitter?

**Comment by MikeMiller567** [comment]
Let's say that it is a recent Neoverse core.

I played around with the parser a little and it's a bit more complex than I originally thought, because decimal numbers as the first token in the line are assumed to be time, and some hex numbers look like decimal numbers, so we get false positives on the "detect time" code in parser.cc (around line 285):

```
        // Tarmac lines often, but not always, start with a timestamp.
        Time time;
        if (tok.isdecimal()) {
```

For my _particular_ emitter, I always have the "ps", so I can put in a hack like this:

```
if (tok.isdecimal() and (line.find("ps")!=std::string::npos)) {
```

But my understanding from the code is this strategy won't work for some formats as they don't have any units associated with the time. Any suggestions on this issue would be welcome.

I did a hack on the memory record and it turned out easy, but unpleasant, as I used a `static` variable to track LD versus ST. Around line 925:

```
        } else if (tok == "LD" || tok == "ST" || tok.ishex()) {
            // Diagrammatic memory access event.

            // making the bool static is a really big hack so that we know what
            // a previous memory access was in case the LD/ST prefix isn't on the line...
            static bool read;

            if (tok == "LD" || tok == "ST") {
                read = (tok == "LD");
                tok = lex();
            }
```

I'm **not** suggesting that we use this code as-is, but it is a proof-of-concept that it is able to parse the file successfully with these hacks.

**Comment by MikeMiller567** [comment]
I like your first suggestion better. It seems simpler in terms of keeping the logic together in a small bunch around the time-processing. I did some coding and was also able to eliminate the `static` variable. Let me know what you think. I made the changes minimal.

```
diff --git a/lib/parser.cpp b/lib/parser.cpp
index 4f7c9df..c3e425c 100644
--- a/lib/parser.cpp
+++ b/lib/parser.cpp
@@ -161,6 +161,13 @@ class TarmacLineParserImpl {
     set<string> unrecognised_system_operations_reported;
     set<string> unrecognised_tarmac_events_reported;
     ParseReceiver *receiver;
+    bool timescale_token_expected;
+    bool read; // Indicates if the last memory token seen was an LD versus an ST
+
+    TarmacLineParserImpl()
+        : timescale_token_expected(false)
+    {
+    }
 
     inline bool iswordchr(char c)
     {
@@ -292,6 +299,11 @@ class TarmacLineParserImpl {
                 // Any of these is something we recognise as a unit of
                 // time, so skip over it.
                 tok = lex();
+                timescale_token_expected = true;
+            } else if (timescale_token_expected) {
+                // we expected a timescale, but it isn't here, so this must be a memory op, so reset the line
+                pos = 0;
+                tok = lex();
             }
 
             last_timestamp = time;
@@ -923,12 +935,13 @@ class TarmacLineParserImpl {
 
             MemoryEvent ev(time, read, size, addr, true, contents);
             receiver->got_event(ev);
-        } else if (tok == "LD" || tok == "ST") {
+        } else if (tok == "LD" || tok == "ST" || tok.ishex()) {
             // Diagrammatic memory access event.
 
-            bool read = (tok == "LD");
-            tok = lex();
-
+            if (tok == "LD" || tok == "ST") {
+                read = (tok == "LD");
+                tok = lex();
+            }
             // Expect a hex address.
             if (!tok.ishex())
                 parse_error(tok, "expected load/store memory address");
```

**Comment by statham-arm** [comment]
Thanks for the suggested patch, and sorry I haven't had a chance to look at this yet!

I was hoping to fix this issue as part of a batch of other parsing issues reported recently, but as it turned out, I had time to fix all the others, and this one is still undone. Worse, for one of the other fixes, I've had to make changes in the area of timestamp handling (issue #13 had a trace snippet in which there was no space between the timestamp and the scale unit). So this patch of yours won't apply cleanly any more.

I quite like the general idea, though. You haven't done quite what I suggested in my option 1, which was to remember the _specific_ time scale unit. But I prefer what you've done instead, because it's more stateless!

The only thing I'd suggest changing is to restrict this special case to _only_ situations where the previous line we saw was `LD` or `ST`, so that in any other circumstance, we'll _know_ we don't understand what we're seeing.

(Also, I'm going to be on holiday for several weeks now, so if you do feel like preparing a revised patch, there's no need to hurry about it.)
